File: sdk/extension-sdk-ts/test/host_api.test.ts
import { describe, expect, it } from "vitest";

import { Extension } from "../src/extension.js";
import { RPCError } from "../src/transport.js";
import { CAPABILITIES, HOOKS, PROTOCOL_VERSION } from "../src/types.js";
import { TestHarness } from "../src/testing/test_harness.js";
import { createMockTransportPair } from "../src/testing/mock_transport.js";

describe("HostAPI", () => {
  it("round-trips host.tasks.create through the test harness", async () => {
    const extension = new Extension("sdk-ext", "1.0.0").withCapabilities(CAPABILITIES.tasksCreate);
    const harness = new TestHarness({
      granted_capabilities: [CAPABILITIES.tasksCreate],
    });
    harness.handleHostMethod("host.tasks.create", async params => {
      expect(params).toEqual({
        workflow: "demo",
        title: "Hello",
        body: "Body",
        frontmatter: { status: "pending", type: "docs" },
      });
      return {
        workflow: "demo",
        number: 7,
        path: ".rc/tasks/demo/task_07.md",
        status: "pending",
      };
    });

    const runPromise = harness.run(extension);
    await harness.initialize({
      name: "sdk-ext",
      version: "1.0.0",
      source: "workspace",
    });

    await expect(
      extension.host.tasks.create({
        workflow: "demo",
        title: "Hello",
        body: "Body",
        frontmatter: { status: "pending", type: "docs" },
      })
    ).resolves.toMatchObject({ number: 7 });

    await harness.shutdown({ reason: "run_completed", deadline_ms: 1000 });
    await expect(runPromise).resolves.toBeUndefined();
  });

  it("returns a run id from host.runs.start", async () => {
    const extension = new Extension("sdk-ext", "1.0.0").withCapabilities(CAPABILITIES.runsStart);
    const harness = new TestHarness({
      granted_capabilities: [CAPABILITIES.runsStart],
    });
    harness.handleHostMethod("host.runs.start", async () => ({
      run_id: "run-child-001",
      parent_run_id: "run-parent-001",
    }));

    const runPromise = harness.run(extension);
    await harness.initialize({
      name: "sdk-ext",
      version: "1.0.0",
      source: "workspace",
    });

    await expect(
      extension.host.runs.start({
        runtime: { name: "child" },
      })
    ).resolves.toEqual({
      run_id: "run-child-001",
      parent_run_id: "run-parent-001",
    });

    await harness.shutdown({ reason: "run_completed", deadline_ms: 1000 });
    await expect(runPromise).resolves.toBeUndefined();
  });

  it("returns absent workflow memory documents", async () => {
    const extension = new Extension("sdk-ext", "1.0.0").withCapabilities(CAPABILITIES.memoryRead);
    const harness = new TestHarness({
      granted_capabilities: [CAPABILITIES.memoryRead],
    });
    harness.handleHostMethod("host.memory.read", async () => ({
      path: ".rc/tasks/demo/memory/task_03.md",
      content: "",
      exists: false,
      needs_compaction: false,
    }));

    const runPromise = harness.run(extension);
    await harness.initialize({
      name: "sdk-ext",
      version: "1.0.0",
      source: "workspace",
    });

    await expect(
      extension.host.memory.read({
        workflow: "demo",
        task_file: "task_03.md",
      })
    ).resolves.toEqual({
      path: ".rc/tasks/demo/memory/task_03.md",
      content: "",
      exists: false,
      needs_compaction: false,
    });

    await harness.shutdown({ reason: "run_completed", deadline_ms: 1000 });
    await expect(runPromise).resolves.toBeUndefined();
  });

  it("correlates out-of-order host responses by id", async () => {
    const extension = new Extension("sdk-ext", "1.0.0").withCapabilities(
      CAPABILITIES.tasksRead,
      CAPABILITIES.memoryRead
    );
    const [extensionTransport, hostTransport] = createMockTransportPair();
    const runPromise = extension.withTransport(extensionTransport).start();

    await hostTransport.writeMessage({
      id: 1,
      method: "initialize",
      params: {
        protocol_version: PROTOCOL_VERSION,
        supported_protocol_versions: [PROTOCOL_VERSION],
        rc_version: "dev",
        extension: {
          name: "sdk-ext",
          version: "1.0.0",
          source: "workspace",
        },
        granted_capabilities: [CAPABILITIES.tasksRead, CAPABILITIES.memoryRead],
        runtime: {
          run_id: "run-1",
          workspace_root: ".",
          invoking_command: "tasks run",
          shutdown_timeout_ms: 1000,
          default_hook_timeout_ms: 5000,
        },
      },
    });

    const initializeResponse = await hostTransport.readMessage();
    expect(initializeResponse.result).toMatchObject({
      protocol_version: PROTOCOL_VERSION,
    });

    const tasksPromise = extension.host.tasks.list({ workflow: "demo" });
    const memoryPromise = extension.host.memory.read({ workflow: "demo" });

    const first = await hostTransport.readMessage();
    const second = await hostTransport.readMessage();

    const memoryRequest = first.method === "host.memory.read" ? first : second;
    const taskRequest = first.method === "host.tasks.list" ? first : second;

    await hostTransport.writeMessage({
      id: memoryRequest.id,
      result: {
        path: ".rc/tasks/demo/memory/MEMORY.md",
        content: "",
        exists: false,
        needs_compaction: false,
      },
    });
    await hostTransport.writeMessage({
      id: taskRequest.id,
      result: [{ workflow: "demo", number: 1, path: "task_01.md", status: "pending" }],
    });

    await expect(memoryPromise).resolves.toMatchObject({ exists: false });
    await expect(tasksPromise).resolves.toMatchObject([{ number: 1 }]);

    await hostTransport.writeMessage({
      id: 2,
      method: "shutdown",
      params: { reason: "run_completed", deadline_ms: 1000 },
    });
    const shutdownResponse = await hostTransport.readMessage();
    expect(shutdownResponse.result).toEqual({ acknowledged: true });

    await expect(runPromise).resolves.toBeUndefined();
  });

  it("returns structured errors from the host", async () => {
    const extension = new Extension("sdk-ext", "1.0.0").withCapabilities(
      CAPABILITIES.artifactsWrite
    );
    const harness = new TestHarness({
      granted_capabilities: [CAPABILITIES.artifactsWrite],
    });
    harness.handleHostMethod("host.artifacts.write", async () => {
      throw new RPCError(-32001, "Capability denied", { reason: "path_out_of_scope" });
    });

    const runPromise = harness.run(extension);
    await harness.initialize({
      name: "sdk-ext",
      version: "1.0.0",
      source: "workspace",
    });

    await expect(
      extension.host.artifacts.write({
        path: "../bad.txt",
        content: "bad",
      })
    ).rejects.toMatchObject<RPCError>({
      code: -32001,
      data: { reason: "path_out_of_scope" },
    });

    await harness.shutdown({ reason: "run_completed", deadline_ms: 1000 });
    await expect(runPromise).resolves.toBeUndefined();
  });

  it("rejects pending calls when the harness transport terminates", async () => {
    const extension = new Extension("sdk-ext", "1.0.0").onPromptPostBuild(
      () => new Promise<never>(() => {})
    );
    const harness = new TestHarness({
      granted_capabilities: [CAPABILITIES.promptMutate],
    });

    void harness.run(extension).catch(() => undefined);
    await harness.initialize({
      name: "sdk-ext",
      version: "1.0.0",
      source: "workspace",
    });

    const pending = harness.dispatchHook(
      "hook-pending",
      {
        name: HOOKS.promptPostBuild,
        event: HOOKS.promptPostBuild,
        mutable: true,
        required: false,
        priority: 500,
        timeout_ms: 5000,
      },
      { prompt_text: "original" }
    );

    await harness.extensionTransport.close();

    await expect(pending).rejects.toThrow("test harness terminated");
  });
});

File: sdk/extension/host_api_test.go
package extension_test

import (
	"context"
	"encoding/json"
	"testing"
	"time"

	extension "github.com/rodolfochicone/rc-project/sdk/extension"
	exttesting "github.com/rodolfochicone/rc-project/sdk/extension/testing"
)

func TestHostAPITasksCreateRoundTripsMethodAndParams(t *testing.T) {
	t.Parallel()

	const name = "sdk-ext"
	const version = "1.0.0"
	ext := extension.New(name, version).WithCapabilities(extension.CapabilityTasksCreate)
	harness, ctx, cancel, errCh := runHarnessedExtension(t, ext, exttesting.HarnessOptions{
		GrantedCapabilities: []extension.Capability{extension.CapabilityTasksCreate},
	})
	defer cancel()

	harness.HandleHostMethod("host.tasks.create", func(_ context.Context, params json.RawMessage) (any, error) {
		var req extension.TaskCreateRequest
		if err := json.Unmarshal(params, &req); err != nil {
			t.Fatalf("unmarshal params: %v", err)
		}
		if req.Workflow != "demo" || req.Title != "Hello" {
			t.Fatalf("unexpected create request: %#v", req)
		}
		if !req.UpdateIndex {
			t.Fatalf("req.UpdateIndex = false, want true")
		}
		return extension.Task{
			Workflow: "demo",
			Number:   7,
			Path:     ".rc/tasks/demo/task_07.md",
			Status:   "pending",
		}, nil
	})

	if _, err := harness.Initialize(ctx, extension.InitializeRequestIdentity{
		Name:    name,
		Version: version,
		Source:  "workspace",
	}); err != nil {
		t.Fatalf("Initialize() error = %v", err)
	}

	task, err := ext.Host().Tasks.Create(ctx, extension.TaskCreateRequest{
		Workflow: "demo",
		Title:    "Hello",
		Body:     "Body",
		Frontmatter: extension.TaskFrontmatter{
			Status: "pending",
			Type:   "backend",
		},
		UpdateIndex: true,
	})
	if err != nil {
		t.Fatalf("Tasks.Create() error = %v", err)
	}
	if task.Number != 7 {
		t.Fatalf("task.Number = %d, want 7", task.Number)
	}

	waitForHostMethod(t, harness, "host.tasks.create")
	shutdownHarness(ctx, t, harness, errCh)
}

func TestHostAPIRunsStartReturnsRunHandle(t *testing.T) {
	t.Parallel()

	const name = "sdk-ext"
	const version = "1.0.0"
	ext := extension.New(name, version).WithCapabilities(extension.CapabilityRunsStart)
	harness, ctx, cancel, errCh := runHarnessedExtension(t, ext, exttesting.HarnessOptions{
		GrantedCapabilities: []extension.Capability{extension.CapabilityRunsStart},
	})
	defer cancel()

	harness.HandleHostMethod("host.runs.start", func(_ context.Context, _ json.RawMessage) (any, error) {
		return extension.RunHandle{
			RunID:       "run-child-001",
			ParentRunID: "run-parent-001",
		}, nil
	})

	if _, err := harness.Initialize(ctx, extension.InitializeRequestIdentity{
		Name:    name,
		Version: version,
		Source:  "workspace",
	}); err != nil {
		t.Fatalf("Initialize() error = %v", err)
	}

	handle, err := ext.Host().Runs.Start(ctx, extension.RunStartRequest{
		Runtime: extension.RunConfig{Name: "child"},
	})
	if err != nil {
		t.Fatalf("Runs.Start() error = %v", err)
	}
	if handle.RunID != "run-child-001" || handle.ParentRunID != "run-parent-001" {
		t.Fatalf("handle = %#v, want run ids", handle)
	}

	shutdownHarness(ctx, t, harness, errCh)
}

func TestHostAPIMemoryReadReturnsAbsentDocument(t *testing.T) {
	t.Parallel()

	const name = "sdk-ext"
	const version = "1.0.0"
	ext := extension.New(name, version).WithCapabilities(extension.CapabilityMemoryRead)
	harness, ctx, cancel, errCh := runHarnessedExtension(t, ext, exttesting.HarnessOptions{
		GrantedCapabilities: []extension.Capability{extension.CapabilityMemoryRead},
	})
	defer cancel()

	harness.HandleHostMethod("host.memory.read", func(_ context.Context, _ json.RawMessage) (any, error) {
		return extension.MemoryReadResult{
			Path:            ".rc/tasks/demo/memory/task_03.md",
			Content:         "",
			Exists:          false,
			NeedsCompaction: false,
		}, nil
	})

	if _, err := harness.Initialize(ctx, extension.InitializeRequestIdentity{
		Name:    name,
		Version: version,
		Source:  "workspace",
	}); err != nil {
		t.Fatalf("Initialize() error = %v", err)
	}

	result, err := ext.Host().Memory.Read(ctx, extension.MemoryReadRequest{
		Workflow: "demo",
		TaskFile: "task_03.md",
	})
	if err != nil {
		t.Fatalf("Memory.Read() error = %v", err)
	}
	if result.Exists || result.Content != "" {
		t.Fatalf("result = %#v, want exists=false content=\"\"", result)
	}

	shutdownHarness(ctx, t, harness, errCh)
}

func TestHostAPIArtifactsWriteReturnsPathOutOfScope(t *testing.T) {
	t.Parallel()

	const name = "sdk-ext"
	const version = "1.0.0"
	ext := extension.New(name, version).WithCapabilities(extension.CapabilityArtifactsWrite)
	harness, ctx, cancel, errCh := runHarnessedExtension(t, ext, exttesting.HarnessOptions{
		GrantedCapabilities: []extension.Capability{extension.CapabilityArtifactsWrite},
	})
	defer cancel()

	harness.HandleHostMethod("host.artifacts.write", func(_ context.Context, _ json.RawMessage) (any, error) {
		return nil, &extension.Error{
			Code:    -32001,
			Message: "Capability denied",
			Data:    extension.MustJSON(map[string]any{"reason": "path_out_of_scope"}),
		}
	})

	if _, err := harness.Initialize(ctx, extension.InitializeRequestIdentity{
		Name:    name,
		Version: version,
		Source:  "workspace",
	}); err != nil {
		t.Fatalf("Initialize() error = %v", err)
	}

	_, err := ext.Host().Artifacts.Write(ctx, extension.ArtifactWriteRequest{
		Path:    "../bad.txt",
		Content: []byte("nope"),
	})
	requestErr := assertRPCErrorCode(t, err, -32001)
	var data struct {
		Reason string `json:"reason"`
	}
	if err := requestErr.DecodeData(&data); err != nil {
		t.Fatalf("DecodeData() error = %v", err)
	}
	if data.Reason != "path_out_of_scope" {
		t.Fatalf("reason = %q, want path_out_of_scope", data.Reason)
	}

	shutdownHarness(ctx, t, harness, errCh)
}

func TestHostAPICallCorrelatesOutOfOrderResponses(t *testing.T) {
	t.Parallel()

	const name = "sdk-ext"
	const version = "1.0.0"
	ext := extension.New(name, version).
		WithCapabilities(extension.CapabilityTasksRead, extension.CapabilityMemoryRead)
	extensionSide, hostSide := exttesting.NewMockTransportPair()
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	errCh := make(chan error, 1)
	go func() {
		errCh <- ext.WithTransport(extensionSide).Start(ctx)
	}()

	if err := hostSide.WriteMessage(extension.Message{
		ID:     json.RawMessage("1"),
		Method: "initialize",
		Params: extension.MustJSON(extension.InitializeRequest{
			ProtocolVersion:           extension.ProtocolVersion,
			SupportedProtocolVersions: []string{extension.ProtocolVersion},
			RcVersion:                 "dev",
			Extension: extension.InitializeRequestIdentity{
				Name:    name,
				Version: version,
				Source:  "workspace",
			},
			GrantedCapabilities: []extension.Capability{
				extension.CapabilityTasksRead,
				extension.CapabilityMemoryRead,
			},
			Runtime: extension.InitializeRuntime{
				RunID:                "run-001",
				WorkspaceRoot:        ".",
				InvokingCommand:      "start",
				ShutdownTimeoutMS:    1000,
				DefaultHookTimeoutMS: 5000,
			},
		}),
	}); err != nil {
		t.Fatalf("write initialize: %v", err)
	}

	initializeResponse, err := hostSide.Receive(ctx)
	if err != nil {
		t.Fatalf("receive initialize response: %v", err)
	}
	if initializeResponse.Error != nil {
		t.Fatalf("initialize error = %v", initializeResponse.Error)
	}

	taskCh := make(chan error, 1)
	memCh := make(chan error, 1)
	var (
		listResult []extension.Task
		memResult  *extension.MemoryReadResult
	)

	go func() {
		result, err := ext.Host().Tasks.List(ctx, extension.TaskListRequest{Workflow: "demo"})
		listResult = result
		taskCh <- err
	}()
	go func() {
		result, err := ext.Host().Memory.Read(ctx, extension.MemoryReadRequest{Workflow: "demo"})
		memResult = result
		memCh <- err
	}()

	first, err := hostSide.Receive(ctx)
	if err != nil {
		t.Fatalf("receive first host call: %v", err)
	}
	second, err := hostSide.Receive(ctx)
	if err != nil {
		t.Fatalf("receive second host call: %v", err)
	}
	if first.Method == second.Method {
		t.Fatalf("expected distinct methods, got %q and %q", first.Method, second.Method)
	}

	requests := map[string]json.RawMessage{
		first.Method:  append(json.RawMessage(nil), first.ID...),
		second.Method: append(json.RawMessage(nil), second.ID...),
	}

	if err := hostSide.WriteMessage(extension.Message{
		ID: requests["host.memory.read"],
		Result: extension.MustJSON(
			extension.MemoryReadResult{Path: ".rc/tasks/demo/memory/MEMORY.md", Exists: false},
		),
	}); err != nil {
		t.Fatalf("write second response: %v", err)
	}
	if err := hostSide.WriteMessage(extension.Message{
		ID:     requests["host.tasks.list"],
		Result: extension.MustJSON([]extension.Task{{Workflow: "demo", Number: 1}}),
	}); err != nil {
		t.Fatalf("write first response: %v", err)
	}

	if err := <-taskCh; err != nil {
		t.Fatalf("Tasks.List() error = %v", err)
	}
	if err := <-memCh; err != nil {
		t.Fatalf("Memory.Read() error = %v", err)
	}
	if len(listResult) != 1 || listResult[0].Number != 1 {
		t.Fatalf("listResult = %#v, want one task", listResult)
	}
	if memResult == nil || memResult.Exists {
		t.Fatalf("memResult = %#v, want exists=false", memResult)
	}

	if err := hostSide.WriteMessage(extension.Message{
		ID:     json.RawMessage("99"),
		Method: "shutdown",
		Params: extension.MustJSON(extension.ShutdownRequest{Reason: "run_completed", DeadlineMS: 1000}),
	}); err != nil {
		t.Fatalf("write shutdown: %v", err)
	}
	shutdownResponse, err := hostSide.Receive(ctx)
	if err != nil {
		t.Fatalf("receive shutdown response: %v", err)
	}
	if shutdownResponse.Error != nil {
		t.Fatalf("shutdown error = %v", shutdownResponse.Error)
	}

	select {
	case err := <-errCh:
		if err != nil {
			t.Fatalf("Start() terminal error = %v", err)
		}
	case <-time.After(2 * time.Second):
		t.Fatal("timed out waiting for extension shutdown")
	}
}

func TestHostAPICallBeforeInitializeFails(t *testing.T) {
	t.Parallel()

	ext := extension.New("sdk-ext", "1.0.0")
	_, err := ext.Host().Tasks.Create(context.Background(), extension.TaskCreateRequest{})
	assertRPCErrorCode(t, err, -32003)
}

func TestHostAPICallWithoutAcceptedCapabilityFails(t *testing.T) {
	t.Parallel()

	const name = "sdk-ext"
	const version = "1.0.0"
	ext := extension.New(name, version)
	harness, ctx, cancel, errCh := runHarnessedExtension(t, ext, exttesting.HarnessOptions{})
	defer cancel()

	if _, err := harness.Initialize(ctx, extension.InitializeRequestIdentity{
		Name:    name,
		Version: version,
		Source:  "workspace",
	}); err != nil {
		t.Fatalf("Initialize() error = %v", err)
	}

	_, err := ext.Host().Tasks.Create(ctx, extension.TaskCreateRequest{})
	assertRPCErrorCode(t, err, -32001)

	shutdownHarness(ctx, t, harness, errCh)
}
